File: castvibe/_connection.py
"""Per-client Cast connection handler.

Wraps an asyncio reader/writer pair and dispatches incoming Cast messages
by namespace.  Device authentication and heartbeat are handled locally;
all other messages are forwarded to an optional callback.
"""

from __future__ import annotations

import asyncio
import contextlib
import json
import logging
from typing import TYPE_CHECKING, Any

import castvibe._namespace as ns
from castvibe._auth import build_auth_response
from castvibe._framing import FramingError, read_message, write_message
from castvibe._log import get_logger
from castvibe._proto.cast_channel_pb2 import (
    CastMessage,
    DeviceAuthMessage,
    HashAlgorithm,
    SignatureAlgorithm,
)

if TYPE_CHECKING:
    from collections.abc import Awaitable, Callable

    from castvibe._certificate import CertificateBundle

log = get_logger("connection")


class Connection:
    """A single Cast sender connection.

    Reads length-prefixed protobuf frames, dispatches by namespace:

    * **deviceauth** — binary auth challenge → auth response (handled locally)
    * **heartbeat** — PING → PONG (handled locally)
    * **everything else** — forwarded to *on_message* callback

    Parameters
    ----------
    reader:
        The asyncio stream reader for this connection.
    writer:
        The asyncio stream writer for this connection.
    bundle:
        Certificate material for device authentication responses.
    on_message:
        Optional async callback ``(connection, cast_message) -> None``
        invoked for every message not handled locally.
    on_disconnect:
        Optional async callback ``(connection) -> None`` invoked when
        the connection is closed (cleanly or due to error).
    """

    __slots__ = (
        "_bundle",
        "_crl",
        "_on_disconnect",
        "_on_message",
        "_reader",
        "_writer",
        "peer",
    )

    def __init__(
        self,
        reader: asyncio.StreamReader,
        writer: asyncio.StreamWriter,
        bundle: CertificateBundle,
        *,
        crl: bytes | None = None,
        on_message: Callable[[Connection, CastMessage], Awaitable[None]] | None = None,
        on_disconnect: Callable[[Connection], Awaitable[None]] | None = None,
    ) -> None:
        self._reader = reader
        self._writer = writer
        self._bundle = bundle
        # CRL is captured at connection-accept time; if CRL rotation is
        # added later, connections should reference the server's CRL instead.
        self._crl = crl
        self._on_message = on_message
        self._on_disconnect = on_disconnect

        # Best-effort peer address for logging.
        peername = writer.get_extra_info("peername")
        self.peer: str = f"{peername[0]}:{peername[1]}" if peername else "unknown"

    # ------------------------------------------------------------------
    # Main loop
    # ------------------------------------------------------------------

    async def handle(self) -> None:
        """Run the message loop until the connection closes.

        This method reads messages in a loop and dispatches them by
        namespace.  It returns when the peer disconnects or an
        unrecoverable framing error is encountered.
        """
        log.info("connection opened: %s", self.peer)
        try:
            await self._loop()
        finally:
            self._writer.close()
            # writer.wait_closed() may raise if the transport is already
            # gone; swallow that so cleanup always completes.
            with contextlib.suppress(OSError):
                await self._writer.wait_closed()
            log.info("connection closed: %s", self.peer)
            if self._on_disconnect is not None:
                await self._on_disconnect(self)

    async def _loop(self) -> None:
        """Read messages until EOF or fatal error."""
        while True:
            try:
                msg = await read_message(self._reader)
            except asyncio.IncompleteReadError:
                # Clean disconnect (EOF).
                return
            except FramingError:
                log.warning("%s: framing error, closing", self.peer, exc_info=True)
                return
            except (ConnectionResetError, BrokenPipeError):
                # Transport-level disconnect.
                return

            try:
                await self._dispatch(msg)
            except Exception:
                # Per-message errors must not kill the connection.
                log.warning(
                    "%s: error handling message on %s",
                    self.peer,
                    msg.namespace,
                    exc_info=True,
                )

    # ------------------------------------------------------------------
    # Dispatch
    # ------------------------------------------------------------------

    async def _dispatch(self, msg: CastMessage) -> None:
        """Route *msg* to the correct handler by namespace."""
        # Warn on unexpected protocol version (only one has ever existed).
        if msg.protocol_version != CastMessage.CASTV2_1_0:
            log.warning(
                "%s: unexpected protocol version %s",
                self.peer,
                msg.protocol_version,
            )

        namespace = msg.namespace

        if namespace == ns.DEVICE_AUTH:
            await self._handle_device_auth(msg)
        elif namespace == ns.HEARTBEAT:
            await self._handle_heartbeat(msg)
        else:
            # Log non-heartbeat messages at DEBUG for diagnostics.
            _log_message(msg, self.peer)
            if self._on_message is not None:
                await self._on_message(self, msg)

    # ------------------------------------------------------------------
    # Local handlers
    # ------------------------------------------------------------------

    async def _handle_device_auth(self, msg: CastMessage) -> None:
        """Respond to a device-auth challenge with the pre-computed auth response."""
        challenge = DeviceAuthMessage()
        if not challenge.ParseFromString(msg.payload_binary):
            log.warning("%s: failed to parse device auth challenge", self.peer)
            return

        if challenge.HasField("challenge"):
            challenge_msg = challenge.challenge
            sender_nonce = challenge_msg.sender_nonce
            log.debug(
                "%s: device auth challenge received (hash=%s sig=%s nonce_len=%d)",
                self.peer,
                _enum_name(HashAlgorithm, challenge_msg.hash_algorithm),
                _enum_name(
                    SignatureAlgorithm,
                    challenge_msg.signature_algorithm,
                ),
                len(sender_nonce),
            )
        else:
            log.debug(
                "%s: device auth challenge received (no challenge field)", self.peer
            )

        payload = build_auth_response(self._bundle, crl=self._crl)
        await self.send_binary(
            source_id=msg.destination_id,
            dest_id=msg.source_id,
            namespace=ns.DEVICE_AUTH,
            data=payload,
        )
        response = DeviceAuthMessage()
        _ = response.ParseFromString(payload)
        if response.HasField("response"):
            response_msg = response.response
            log.debug(
                "%s: device auth response sent (hash=%s sig=%s ica=%d crl_len=%d nonce_len=%d)",
                self.peer,
                _enum_name(HashAlgorithm, response_msg.hash_algorithm),
                _enum_name(
                    SignatureAlgorithm,
                    response_msg.signature_algorithm,
                ),
                len(response_msg.intermediate_certificate),
                len(response_msg.crl),
                len(response_msg.sender_nonce),
            )
        else:
            log.debug("%s: device auth response sent", self.peer)

    async def _handle_heartbeat(self, msg: CastMessage) -> None:
        """Respond to PING with PONG (fast-path, no JSON parsing)."""
        if msg.payload_type == CastMessage.STRING and '"PING"' in msg.payload_utf8:
            await self.send_json(
                source_id=msg.destination_id,
                dest_id=msg.source_id,
                namespace=ns.HEARTBEAT,
                data={"type": "PONG"},
            )

    # ------------------------------------------------------------------
    # Sending helpers
    # ------------------------------------------------------------------

    async def send_json(
        self,
        source_id: str,
        dest_id: str,
        namespace: str,
        data: dict[str, Any],
    ) -> None:
        """Send a JSON (STRING) Cast message."""
        msg = _build_message(
            source_id=source_id,
            dest_id=dest_id,
            namespace=namespace,
            payload_type=CastMessage.STRING,
            payload_utf8=json.dumps(data, separators=(",", ":")),
        )
        await write_message(self._writer, msg)

    async def send_binary(
        self,
        source_id: str,
        dest_id: str,
        namespace: str,
        data: bytes,
    ) -> None:
        """Send a binary (BINARY) Cast message."""
        msg = _build_message(
            source_id=source_id,
            dest_id=dest_id,
            namespace=namespace,
            payload_type=CastMessage.BINARY,
            payload_binary=data,
        )
        await write_message(self._writer, msg)


# ------------------------------------------------------------------
# Helpers
# ------------------------------------------------------------------


def _build_message(
    *,
    source_id: str,
    dest_id: str,
    namespace: str,
    payload_type: CastMessage.PayloadType,
    payload_utf8: str = "",
    payload_binary: bytes = b"",
) -> CastMessage:
    """Construct a ``CastMessage`` with all required fields."""
    msg = CastMessage()
    msg.protocol_version = CastMessage.CASTV2_1_0
    msg.source_id = source_id
    msg.destination_id = dest_id
    msg.namespace = namespace
    msg.payload_type = payload_type
    if payload_type == CastMessage.BINARY:
        msg.payload_binary = payload_binary
    else:
        msg.payload_utf8 = payload_utf8
    return msg


def _log_message(msg: CastMessage, peer: str) -> None:
    """Emit a DEBUG log line summarising a Cast message."""
    if not log.isEnabledFor(logging.DEBUG):
        return
    if msg.payload_type == CastMessage.BINARY:
        payload_repr = f"<{len(msg.payload_binary)} bytes>"
    else:
        payload_repr = msg.payload_utf8
    log.debug(
        "%s: ns=%s src=%s dst=%s payload=%s",
        peer,
        msg.namespace,
        msg.source_id,
        msg.destination_id,
        payload_repr,
    )


def _enum_name(enum_type: type[object], value: int) -> str:
    """Best-effort enum name formatting for protobuf debug logs."""
    name_fn = getattr(enum_type, "Name", None)
    if callable(name_fn):
        try:
            return str(name_fn(value))
        except ValueError:
            return f"UNKNOWN({value})"
    return str(value)

File: crates/vibecast-cast/src/connection.rs
//! Per-sender connection handling.
//!
//! Each connection runs a read loop plus a writer task fed by an mpsc channel
//! (the split-and-channel actor pattern), so the device hub can push messages to
//! a sender concurrently with the read loop. Device-auth and heartbeat are
//! answered locally; every other message is forwarded as a [`ServerEvent`].

use std::sync::Arc;

use futures_util::{SinkExt, StreamExt};
use prost::Message as _;
use tokio::io::{AsyncRead, AsyncWrite};
use tokio::sync::mpsc;
use tokio_util::codec::Framed;

use vibecast_proto::cast_channel::cast_message::{PayloadType, ProtocolVersion};
use vibecast_proto::{
    CastCodec, CastMessage, DeviceAuthMessage, FramingError, HashAlgorithm, SignatureAlgorithm,
};
use vibecast_security::{build_auth_error, build_auth_response, AuthErrorType, CertificateBundle};

use crate::error::CastError;
use crate::{message, namespace as ns};

/// Device-auth material captured for the lifetime of a connection.
#[derive(Clone, Debug)]
pub struct AuthMaterial {
    /// The active certificate bundle (device cert, intermediates, signatures).
    pub bundle: CertificateBundle,
    /// Server-level CRL override; when `None`, the bundle's own CRL is used.
    pub crl: Option<Vec<u8>>,
}

/// A cloneable handle for sending messages to one connected sender.
#[derive(Clone, Debug)]
pub struct ConnectionHandle {
    id: u64,
    peer: Arc<str>,
    tx: mpsc::Sender<CastMessage>,
}

impl ConnectionHandle {
    /// Stable, unique connection id.
    #[must_use]
    pub fn id(&self) -> u64 {
        self.id
    }

    /// Best-effort peer address, for logging.
    #[must_use]
    pub fn peer(&self) -> &str {
        &self.peer
    }

    /// Send a pre-built message to this sender.
    pub async fn send(&self, message: CastMessage) -> Result<(), CastError> {
        self.tx.send(message).await.map_err(|_| CastError::Closed)
    }

    /// Send a JSON (STRING) message (compact, no whitespace).
    pub async fn send_json(
        &self,
        source_id: &str,
        dest_id: &str,
        namespace: &str,
        data: &serde_json::Value,
    ) -> Result<(), CastError> {
        let payload = serde_json::to_string(data)?;
        self.send(message::build_string(
            source_id, dest_id, namespace, payload,
        ))
        .await
    }

    /// Send a BINARY message.
    pub async fn send_binary(
        &self,
        source_id: &str,
        dest_id: &str,
        namespace: &str,
        data: Vec<u8>,
    ) -> Result<(), CastError> {
        self.send(message::build_binary(source_id, dest_id, namespace, data))
            .await
    }
}

/// Lifecycle and message events emitted by connections to the server consumer.
#[derive(Debug)]
pub enum ServerEvent {
    /// A sender connected (after TLS handshake).
    Connected(ConnectionHandle),
    /// A non-local message was received from a sender.
    Message {
        /// Handle to reply to the originating sender.
        handle: ConnectionHandle,
        /// The received message.
        message: CastMessage,
    },
    /// A sender disconnected.
    Disconnected {
        /// The connection id that closed.
        id: u64,
        /// The peer address that closed.
        peer: Arc<str>,
    },
}

/// Aborts the wrapped task when dropped, so the writer never outlives the reader.
struct AbortGuard(tokio::task::JoinHandle<()>);

impl Drop for AbortGuard {
    fn drop(&mut self) {
        self.0.abort();
    }
}

/// Run one connection to completion over `stream`.
///
/// Returns when the peer disconnects or a fatal framing error occurs. Emits
/// [`ServerEvent::Connected`] on entry and [`ServerEvent::Disconnected`] on exit.
pub async fn run_connection<S>(
    stream: S,
    id: u64,
    peer: Arc<str>,
    auth: Arc<AuthMaterial>,
    events: mpsc::Sender<ServerEvent>,
) where
    S: AsyncRead + AsyncWrite + Unpin + Send + 'static,
{
    let (mut sink, mut reader) = Framed::new(stream, CastCodec).split();
    let (tx, mut rx) = mpsc::channel::<CastMessage>(64);
    let handle = ConnectionHandle {
        id,
        peer: Arc::clone(&peer),
        tx,
    };

    // Writer task: drains outbound messages onto the wire. Aborted on drop.
    let _writer = AbortGuard(tokio::spawn(async move {
        while let Some(message) = rx.recv().await {
            if sink.send(message).await.is_err() {
                break;
            }
        }
    }));

    tracing::info!(peer = %peer, "connection opened");
    if events
        .send(ServerEvent::Connected(handle.clone()))
        .await
        .is_err()
    {
        return;
    }

    loop {
        match reader.next().await {
            Some(Ok(message)) => {
                if !dispatch(message, &auth, &handle, &events).await {
                    break;
                }
            }
            Some(Err(FramingError::Io(_))) => break, // transport disconnect
            Some(Err(err)) => {
                tracing::warn!(peer = %peer, error = %err, "framing error, closing");
                break;
            }
            None => break, // clean EOF
        }
    }

    tracing::info!(peer = %peer, "connection closed");
    let _ = events.send(ServerEvent::Disconnected { id, peer }).await;
}

/// Route one message. Returns `false` if the connection should shut down.
async fn dispatch(
    message: CastMessage,
    auth: &AuthMaterial,
    handle: &ConnectionHandle,
    events: &mpsc::Sender<ServerEvent>,
) -> bool {
    if message.protocol_version != ProtocolVersion::Castv210 as i32 {
        tracing::warn!(
            peer = %handle.peer,
            version = message.protocol_version,
            "unexpected protocol version"
        );
    }

    match message.namespace.as_str() {
        ns::DEVICE_AUTH => {
            let Some(payload) = device_auth_response(&message, auth) else {
                return true;
            };
            handle
                .send_binary(
                    &message.destination_id,
                    &message.source_id,
                    ns::DEVICE_AUTH,
                    payload,
                )
                .await
                .is_ok()
        }
        ns::HEARTBEAT => {
            if is_ping(&message) {
                let pong = serde_json::json!({ "type": "PONG" });
                return handle
                    .send_json(
                        &message.destination_id,
                        &message.source_id,
                        ns::HEARTBEAT,
                        &pong,
                    )
                    .await
                    .is_ok();
            }
            true
        }
        _ => events
            .send(ServerEvent::Message {
                handle: handle.clone(),
                message,
            })
            .await
            .is_ok(),
    }
}

/// Build the deviceauth reply for a challenge, or `None` if it can't be parsed.
fn device_auth_response(message: &CastMessage, auth: &AuthMaterial) -> Option<Vec<u8>> {
    let payload = message.payload_binary.as_deref().unwrap_or(&[]);
    let challenge = match DeviceAuthMessage::decode(payload) {
        Ok(challenge) => challenge,
        Err(_) => {
            tracing::warn!("failed to parse device auth challenge");
            return None;
        }
    };

    // proto2 defaults: hash=SHA1, signature=RSASSA_PKCS1v15 when unset.
    let (requested_hash, requested_sig) = match &challenge.challenge {
        Some(inner) => (
            inner.hash_algorithm.unwrap_or(HashAlgorithm::Sha1 as i32),
            inner
                .signature_algorithm
                .unwrap_or(SignatureAlgorithm::RsassaPkcs1v15 as i32),
        ),
        None => (
            HashAlgorithm::Sha1 as i32,
            SignatureAlgorithm::RsassaPkcs1v15 as i32,
        ),
    };

    if requested_sig != SignatureAlgorithm::RsassaPkcs1v15 as i32 {
        return Some(build_auth_error(
            AuthErrorType::SignatureAlgorithmUnavailable,
        ));
    }

    match HashAlgorithm::try_from(requested_hash) {
        Ok(hash) => Some(build_auth_response(&auth.bundle, hash, auth.crl.as_deref())),
        Err(_) => Some(build_auth_error(
            AuthErrorType::SignatureAlgorithmUnavailable,
        )),
    }
}

fn is_ping(message: &CastMessage) -> bool {
    message.payload_type == PayloadType::String as i32
        && message
            .payload_utf8
            .as_deref()
            .is_some_and(|payload| payload.contains("\"PING\""))
}
